contracts/metatx/ERC2771Forwarder.sol

```diff
@@ -0,0 +1,278 @@
+
+    /**
+     * @dev Returns `true` if a request is valid for a provided `signature` at the current block.
+     *
+     * A transaction is considered valid when it hasn't expired (deadline is not met), and the signer
+     * matches the `from` parameter of the signed request.
+     *
+     * NOTE: A request may return false here but it optionally reverts in {batchExecute} to prevent
+     * reverts when a batch includes a request that was already frontrunned by another relay. This
+     * behavior is opt-in by using a flag in {executeBatch}.
+     */
+    function verify(ForwardRequestData calldata request) public view virtual returns (bool) {
+        (bool alive, bool signerMatch, ) = _validate(request, nonces(request.from));
+        return alive && signerMatch;
+    }
+
+    /**
+     * @dev Executes a `request` on behalf of `signature`'s signer guaranteeing that the forwarded call
+     * will receive the requested gas and no ETH is left stuck in the contract.
+     */
+    function execute(ForwardRequestData calldata request) public payable virtual returns (bool) {
+        // This check can be before the call because _execute reverts with an invalid request.
+        if (msg.value != request.value) {
+            revert ERC2771ForwarderMismatchedValue(request.value, msg.value);
+        }
+
+        return _execute(request, _useNonce(request.from), true);
+    }
+
+    /**
+     * @dev Batch version of {execute} with optional atomic behavior and refunding.
+     *
+     * The `atomic` parameter indicates whether all requests are executed or the whole
+     * batch reverts if a single one of them is not a valid request as defined by {verify}.
+     *
+     * The `refundReceiver` is used for optionally send the funds back to a selected address in
+     * case there's ETH left in the contract at the end of the execution instead of reverting. If
+     * the provided refund receiver is the `address(0)`, the contract will revert with an
+     * {ERC2771ForwarderMismatchedValue} error.
+     *
+     * NOTE: The `atomic` flag guarantees an all-or-nothing behavior only for the first level forwarded
+     * calls. In case a call is forwarded to another contract, it may revert without the top-level call
+     * reverting.
+     */
+    function executeBatch(
+        ForwardRequestData[] calldata requests,
+        bool atomic,
+        address payable refundReceiver
+    ) public payable virtual {
+        uint256 requestsValue;
+        uint256 spentValue;
+
+        for (uint256 i; i < requests.length; ++i) {
+            requestsValue += requests[i].value;
+            bool success = _execute(requests[i], _useNonce(requests[i].from), atomic);
+            if (success) {
+                // Will never reallistically overflow given _execute will natively revert at
+                // some point due to insufficient balance.
+                unchecked {
+                    spentValue += requests[i].value;
+                }
```

Comment:
```suggestion
            bool success = _execute(requests[i], _useNonce(requests[i].from), atomic);
            unchecked {
                // Will never reallistically overflow given _execute will natively revert at
                // some point due to insufficient balance.
                requestsValue += requests[i].value;
                spentValue += success ? requests[i].value : 0;
            }
```

Reply:
No. Putting `requestsValue` in an unchecked block allows a relayer to send a signed invalid request (skipped) with `type(uint256).max` and overflow the `requestsValue`.